polaris/polaris/sep10auth/views.py

```diff
@@ -89,20 +90,114 @@ def _transaction_from_post_request(request):
+
+    # make horizon /accounts API call to get signers and threshold
+    account_signers, threshold = _get_signers_and_threshold(source_account)
+
+    ##################################################################
+    # The functionality implemented below should go in stellar_sdk and
+    # invoked like so:
+    #
+    #       verify_challenge_transaction(
+    #           challenge_transaction=envelope_xdr,
+    #           server_account_id=settings.SIGNING_KEY,
+    #           network_passphrase=settings.STELLAR_NETWORK_PASSPHRASE,
+    #           account_signers=account_signers,
+    #           threshold=threshold
+    #       )
+    #
+    # Where account_signers and threshold are defined like so:
+    #
+    #       account_json = server.accounts().account_id(source_account).call()
+    #       account_signers = account_json["signers"]
+    #       threshold = account_json["thresholds"]["med_threshold"]
+    #
+    # This way, the client can choose which threshold the challenge should
+    # meet. It also makes sense to pass the unaltered account_json["signers"]
+    # instead of requiring the client to do some manipulation of the
+    # response before passing.
+    ##################################################################
+
+    # Get the account signers used for this transaction
+    server_kp = Keypair.from_public_key(settings.SIGNING_KEY)
+    matched_signers = []
+    for dec_signer in transaction_envelope.signatures:
+        try:
+            server_kp.verify(tx_hash, dec_signer.signature)
+        except BadSignatureError:
+            # dec_signer is not from the server's keypair
+            matched_signers.append(_match_signer(tx_hash, dec_signer, account_signers))
+
+    # Check threshold
+    if sum(signer["weight"] for signer in matched_signers) < threshold:
+        raise ValueError(
+            "Transaction signers do not reach medium threshold for account"
+        )
+
+
+def _match_signer(tx_hash, dec_signer, account_signers):
+    """
+    Iterate over account_signers to find a match for dec_signer. If
+    dec_signer doesn't have a match, raise a ValueError.
+    """
+    matched_signer = None
+    for acc_signer in account_signers:
+        account_kp = acc_signer["keypair"]
+        if dec_signer.hint != account_kp.signature_hint():
+            continue
+        try:
+            account_kp.verify(tx_hash, dec_signer.signature)
+        except BadSignatureError:
+            continue
+        else:
+            matched_signer = acc_signer
+            break
+
+    if not matched_signer:
+        raise ValueError("Transaction has unrecognized signatures")
+
+    return matched_signer
+
+
+def _get_signers_and_threshold(source_account):
+    """
+    Makes a Horizon API call to /accounts and returns the signers and
+    threshold. Replaces 'key' with 'keypair'.
+    """
+    server = settings.HORIZON_SERVER
+    account_json = server.accounts().account_id(source_account).call()
```

Comment:
❗️ We also need to handle the case that an account does not exist. I don't see that being handled here. If an account doesn't exist we can run with the master key since an account when first created has the master key as its signer and that master key must sign a transaction to make anyone else a signer. https://github.com/stellar/stellar-protocol/pull/489/files#diff-72480d0894ae2d1a31c229b64ad4c37bR135-R137

Reply:
Ok, initially I thought it was a bit weird to authenticate an account that doesn't actually exist, since the eventual deposit would when it tries to submit the payment transaction. But I see [SEP-24](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0024.md#stellar-account-does-not-exist) asks anchors to create the account if it doesn't exist, which we do in `create_stellar_deposit`.

I'll implement the case where it doesn't exist.

Reply:
`verify_challenge_transaction()` already checks that the source account is a signer so if the account doesn't exist, we'll just call the function without `account_signers`.